lib/msf/core/payload/adapter/fetch.rb

```diff
@@ -0,0 +1,130 @@
+        Msf::OptString.new('FETCH_FILENAME', [ false, 'Name to use on remote system when storing payload', Rex::Text.rand_text_alpha(rand(8..12))])
+      ]
+    )
+    register_advanced_options(
+      [
+        Msf::OptAddress.new('FetchListenerBindAddress', [ false, 'The specific IP address to bind to to serve the payload if different from FETCH_SRVHOST']),
+        Msf::OptPort.new('FetchListenerBindPort', [false, 'The port to bind to if different from FETCH_SRVPORT']),
+        Msf::OptBool.new('FetchHandlerDisable', [true, 'Disable fetch handler', false]),
+        Msf::OptString.new('FetchServerName', [true, 'Fetch Server Name', 'Apache']),
+      ]
+    )
+    @fetch_service = nil
+  end
+
+  # copied from https://github.com/rapid7/metasploit-framework/blob/master/lib/msf/core/exploit/remote/socket_server.rb
+  def _determine_server_comm(ip, srv_comm = datastore['ListenerComm'].to_s)
+    comm = nil
+
+    case srv_comm
+    when 'local'
+      comm = ::Rex::Socket::Comm::Local
+    when /\A-?[0-9]+\Z/
+      comm = framework.sessions.get(srv_comm.to_i)
+      raise(RuntimeError, "Socket Server Comm (Session #{srv_comm}) does not exist") unless comm
+      raise(RuntimeError, "Socket Server Comm (Session #{srv_comm}) does not implement Rex::Socket::Comm") unless comm.is_a? ::Rex::Socket::Comm
+    when nil, ''
+      unless ip.nil?
+        comm = Rex::Socket::SwitchBoard.best_comm(ip)
+      end
+    else
+      raise(RuntimeError, "SocketServer Comm '#{srv_comm}' is invalid")
+    end
+
+    comm || ::Rex::Socket::Comm::Local
+  end
+
+  def srvhost
+    datastore['FETCH_SRVHOST']
+  end
+
+  def srvport
+    datastore['FETCH_SRVPORT']
+  end
+
+  def fetch_bindhost
+    datastore['FetchListenerBindAddress'].blank? ? srvhost : datastore['FetchListenerBindAddress']
+  end
+
+  def fetch_bindport
+    datastore['FetchListenerBindPort'].blank? ? srvport : datastore['FetchListenerBindPort']
+  end
+
+  # If no fetch URL is provided, we generate one based off the underlying payload data
+  # This is because if we use a randomly-generated URI, the URI generated by venom and
+  # Framework will not match.  This way, we can build a payload in venom and a listener
+  # in Framework, and if the underlying payload type/host/port are the same, the URI
+  # will be, too.
+  #
+  def default_srvuri
+    # If we're in framework, payload is in datastore; msfvenom has it in refname
+    payload_name = datastore['payload'] ||= refname
+    decoded_uri = payload_name.dup
+    # there may be no transport, so leave the connection string off if that's the case
+    if module_info['ConnectionType'].upcase == 'REVERSE'
+      decoded_uri << datastore['LHOST']
+      decoded_uri << datastore['LPORT'].to_s
+    end
+    if module_info['ConnectionType'].upcase == 'BIND'
+      decoded_uri << datastore['LHOST']
+      decoded_uri << datastore['RPORT'].to_s
+    end
+    Rex::Text.encode_base64url(decoded_uri)[0, 12]
```

Comment:
It turns out the base64 encoded name was waaaaay longer than I think we wanted.  In the even someone has to copy it over, I figured we should keep it relatively shorter.  Open to ideas on the definition of "shorter"